clientconn.go

```diff
@@ -685,14 +686,7 @@ func (ac *addrConn) wait(ctx context.Context, failFast bool) (transport.ClientTr
 // tight loop.
 func (ac *addrConn) tearDown(err error) {
 	ac.mu.Lock()
-	defer func() {
-		ac.mu.Unlock()
-		ac.cc.mu.Lock()
-		if ac.cc.conns != nil {
-			delete(ac.cc.conns, ac.addr)
-		}
-		ac.cc.mu.Unlock()
-	}()
+	defer ac.mu.Unlock()
```

Comment:
I would suggest that still keeping "delete" into this function (use a input bool param to control it), which makes the code easier to maintain.